### ISSUE TITLE
Replace IDE

### ISSUE THREAD

**Issue by LADSoft** [LADSoft/OrangeC#439]
I've been thinking of replacing the IDE with something like [Code::Blocks](http://www.codeblocks.org).   At the cost of writing some orange c specific plugins, it would get this project out of the GUI business.

The milestone that does this would also be likely to correct some of the debugger-related issues, such as #271 (debugging without an IDE) and #252 (PDB File support).

**Comment by chuggafan** [comment]
If I may make a suggestion, making a plugin that just supports the LSP (Language Server Protocol) https://docs.microsoft.com/en-us/visualstudio/extensibility/language-server-protocol?view=vs-2019 and the Debug Adapter Protocol https://microsoft.github.io/debug-adapter-protocol/ would be much more portable since it's actually "standard" and would allow for an easy-to-make VSCode plugin or Visual Studio. https://langserver.org/ shows how many LSP implementors and consumers there are, and I think it makes more sense to make something to a standard like LSP (which has quite a lot of implementors [as seen here](https://microsoft.github.io/language-server-protocol/implementors/servers/) ) then to do our own thing and work poorly because of it... LSP has the nice addition in this case of directly integrating with VS and VSCode so we require very little changes between any editor we want to support. Clang has their own VSCode plugin for their LSP adaptor and it's tiny [as seen here](https://github.com/llvm-mirror/clang-tools-extra/tree/master/clangd/clients/clangd-vscode) so... if we're going that route I vote for LSP and DAG implementations just for our sanity.

**Comment by LADSoft** [comment]
Sure that would work :smile:

**Comment by LADSoft** [comment]
Code::Blocks already does code completion and stuff like that.   I didn't plan to do any additional work for any features it might be missing but rather just let it go...   I assume that means we wouldn't need to explicitly implement LSP?

**Comment by chuggafan** [comment]
The reason I suggested it was to allow for other things than Code::Blocks, and while Code::Blocks seems to have some code completion capabilities Code::Blocks seems to rely exclusively on GCC for everything compiler based and would require some work to change this. And while Code::Blocks might (might) be easier to implementing for now, I think that implementing an LSP that allows for people to use the debugger and code completion from the actual compiler you're using (along with all the errors in source locations) would be better in the long run as you're no longer bound to a single standardized API that's relatively stable.

**Comment by LADSoft** [comment]
ok that sounds interesting....  but rather than develop an entire LSP I might be interested in taking something like [cquery](https://github.com/ladsoft/orangec/issues) and adding the OCC diagnostics.    This might also go along with #245 which I had planned to finish this milestone...

**Comment by chuggafan** [comment]
Adding in CQuery support for the compiler would, in my view, be the same thing as CQuery is basically just a LSP wrapper for libclang. Either way it'd accomplish an LSP that can be used from VSCode, VS, etc.

**Comment by GitMensch** [comment]
I don't feel like I've that much contributed to OrangeC to have a voting strength like you two, but if you have a pat my "from all of these I'd suggest the full LSP route" note may have an influence :-)

Note: Especially as the current samples are all typescript: if you go the "LSP route" please produce a concise "How OrangeC implemented LSP" documentation (which may include some automated JSON creation or whatever).
This will help others doing something similar (_disclaimer: "others" may include me or other GnuCOBOL related persons to implement a COBOL-provider for LSP_).

**Comment by LADSoft** [comment]
I'm still thinking about this...   the tempting thought for me is to just use use clang to do the processing and use cquery as-is (with the exception of adding some orange-c specific code for diagnostics) which is easy to code but that is likely going to drag in too many third-party dependencies and also won't help you with the possibility of doing gnucobol either....    beyond that it is either find something that already implements an interface between the json protocol in use and some kind of 'wrapper' api (gutting cquery may work or there may be something else specifically designed for this) or use jsoncpp and create such a wrapper api from scratch...   in any case such a wrapper is the easy part, there is still all the parsing code that has to be tied into it lol!

**Comment by LADSoft** [comment]
got a request from @GitMensch in #297 to make a new github project to archive OCIDE in.   Basically copy the repository, delete everything but OCIDE and any dependencies, and then push the new project to github.

**Comment by GitMensch** [comment]
Yes, if done from master it should be identical to the last release (I'd still include the CI build but with a "one build" only); after this is done everything OCIDE-only can be removed from master (makefiles and CI builds may need some adjustments) and those remove commits can then be merged to the decoupled branch giving it even more focus and a slightly better time.

**Comment by AlexN-114** [comment]
I think it is a pity that the IDE is no longer part of the Orange C package. The package loses so something of its charm. With it you have a all you need to develop software for Windows (and you can carry it working on an USB-stick). Without it you have a lot of files, whit which you perhaps can build programs and you wouldn't find a debugger.

PS: I think you can change the date in _appveyorversion.bat_ to 2020.

**Comment by chuggafan** [comment]
The IDE is a separate package that can be downloaded separately, the problem with OCIDE is that Compiler and IDE development are both heruclean tasks on their own and maintaining both is nigh-impossible, especially for such a small amount of people. besides by making it so that we can just plug into any IDE it means that it's infinitely easier to deal with later and allows for people (like me) to leverage debugging in those IDEs that are generally superior (VSCode for me). Even microsoft, as big as they are, don't maintain a compiler for their IDE while maintaining the IDE, they pay out to EDG to give them a compiler for their IDE to work with and the compiler team does things separately. 

Even without it, you can still "develop" with this, you just lack a debugger (which if we should switch to using a more up-to-date PDB format with CodeView 8 as an aside). The main problem here is that the OCIDE took up 15 minutes of build time for something that has 1. no tests, 2. almost no changes 3. breaks if you look at it funny, 4. barely has support for debugging it's own projects and has project creation straight out of the early 90's. Maintaining so many things with... 6 total contributors (and Realistically, 1 for most of the time)... is nigh-impossible.

And still, it had to be removed to fix the build time issue, and while making the build take less time IS a goal, at this moment there is very little progress that can be made without major overhauls, so at least for now, to be able to accurately see if builds are passing or failing, it must go.

**Comment by LADSoft** [comment]
yeah I'm kinda sad too...   but truth is very few people seem to use it and I don't really have the inclination to keep trying to maintain it...

While I hadn't planned to remove the IDE immediately it seemed like a good idea to try to resolve the build time issues...

That said there is an upcoming milestone which formalizes interfaces that an IDE/DEBUGGER would use to get symbols and so forth using more or less standard methods.   The plan is when we get there I'll implement that and then find someone else's IDE that uses it to add to the package :smile:

**Comment by GitMensch** [comment]
@Alex: And for now it won't go away, it is just another package with it's own installer (if you want that to use).
Just install / extract both and you're fine and can carry it around on a stick.

**Comment by AlexN-114** [comment]
> @alex: And for now it won't go away, it is just another package with it's own installer (if you want that to use). Just install / extract both and you're fine and can carry it around on a stick.
Thank you! I found it at [https://ci.appveyor.com/project/LADSoft/ocide](url).

**Comment by AlexN-114** [comment]
>  (VSCode for me). 

What is necessary to use VSCode as IDE and Debugger?

**Comment by chuggafan** [comment]
>What is necessary to use VSCode as IDE and Debugger?

Well, for non-OrangeC things the package vscode-cpptools stuff from microsoft. It's what I use for looking because it uses the same intellisense that Visual Studio has. If we actually make an LSP though we can get more exact diagnostics and OrangeC problems seen when I work in VSCode, and eventually we can get debugger support through the same mechanism as vscode-cpptools as there's ways to do that.

**Comment by GitMensch** [comment]
Just to drop it here again: thank you for the [separate OCIDE](https://github.com/LADSoft/OCIDE) package and for including the CI build. Side question: would it be reasonable to have https://scan.coverity.com/projects/ladsoft-ocide (as far as I remember OCIDE was the main "issue owner" there).

I *think* https://github.com/LADSoft/OrangeC/blob/master/doc/general/Ocide.md may be best moved ("integrated") to https://github.com/LADSoft/OCIDE/blob/master/README.md - @LADSoft what is your take on this [the only other thing to change here would be to remove the reference from https://github.com/LADSoft/OrangeC/blob/master/doc/Tools.md]?

**Comment by LADSoft** [comment]
yeah i can move ocide.md.    I'll see about moving the scan for it as well.

**Comment by LADSoft** [comment]
@GitMensch you're welcome
@AlexN-114 there is a plan to rework how the compiler is presented in IDEs in milestone 5.
